docs/specification/draft/basic/transports.mdx

```diff
@@ -287,6 +288,164 @@ protocol version 2024-11-05) as follows:
+
+#### MCP `sessionId` Requirements
+
+- Session IDs **SHOULD** be globally unique and cryptographically secure (e.g., a securely generated UUID, a JWT, or a cryptographic hash)
+- Session IDs **MUST** contain only visible ASCII characters (0x21 to 0x7E)
+- Servers **MUST** maintain exactly one active WebSocket connection per MCP `sessionId`. The most recently established connection becomes the active connection for that MCP session.
+- Upon reconnection, the client calls initialize with a `sessionId` and the server **MUST** close any still open WebSocket connection associated with that session ID with close code `1000` (Normal Closure) and reason "MCP session reconnected on another connection"
+- Close frames might be ignored and there is delay in effectualization, so the server **MUST** also ignore any messages that arrive for the MCP session on any connection but the active one
+- Server-initiated notifications **MUST** be sent over the active connection
+- Clients **SHOULD** keep track of unacknowledged messages and resend them if necessary after reconnection
+
+### WebSocket Transport Lifecycle
+
+#### Initial Connection and Session Establishment
+
+**Client Behavior:**
+- Establish WebSocket connection to server
+- Send MCP `initialize` without `sessionId`
+- Wait for `initialize` response and extract `sessionId`
+- Store `sessionId` for reconnection
+- Send `notifications/initialized` with `sessionId` [TODO: SEP-1364 envisions making this optional. Adjust once finalized]
+
+**Server Behavior:**
+- Accept WebSocket connection
+- Generate and include `sessionId` when responding to `initialize` request
+
+#### Normal Operation
+
+During normal operation:
+- Client **MUST** include the `sessionId` in all messages except the first `initialize`
+- Server **MUST** close the WebSocket connection with close code `1008` (Policy Violation) and reason "Missing sessionId" for requests without `sessionId` (other than `initialize`)
+- Server maintains session state and processes requests normally
+- Server may send notifications to the client at any time
+- Both client and server should handle JSON-RPC requests and responses according to MCP specification
+
+#### Disconnection Handling
+
+WebSocket connection interruption might occur at any time (e.g., due to network conditions). Therefore:
+- WebSocket connection interruption **MUST NOT** be interpreted as the client cancelling its request
+- To cancel, the client **SHOULD** explicitly send an MCP `CancelledNotification`
+- Similarly, disconnection **MUST NOT** be interpreted as terminating the MCP session
+- Server **MUST** preserve session state for a reasonable timeout period to allow reconnection
+
+#### Reconnection and Session Recovery
+
+Upon reconnection:
```

Comment:
Streamable HTTP allows event replay via `Last-Event-Id`. Should we include something similar for websocket reconnections?

Reply:
Once I started thinking about this more, I hit some questions. Notifications don't include a JSON-RPC message and that's probably the most interesting thing that a client would miss. What id is used for Last-Event-Id in the Streamable HTTP transport? I also think that messages are better tied to sessions and this SEP draws a cleaner line between sessions and connections... although the Streamable HTTP transport was starting to do that also.

Thoughts?